File: backend/src/api/audit.py
from datetime import date
from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.src.api.deps import require_auditor
from backend.src.api.schemas import ClearFlagRequest, ClaimResponse
from backend.src.db.session import get_db
from backend.src.services.audit_service import AuditService

router = APIRouter(prefix="/audit", tags=["audit"])

_svc = AuditService()


@router.get("/claims", response_model=list[ClaimResponse])
async def get_flagged_claims(
    violation_type: Optional[str] = Query(None),
    date_from: Optional[date] = Query(None),
    date_to: Optional[date] = Query(None),
    employee_id: Optional[UUID] = Query(None),
    current_user=Depends(require_auditor),
    db: AsyncSession = Depends(get_db),
):
    return await _svc.get_flagged_claims(db, violation_type, date_from, date_to, employee_id)


@router.post("/claims/{claim_id}/investigate")
async def investigate_claim(
    claim_id: UUID,
    current_user=Depends(require_auditor),
    db: AsyncSession = Depends(get_db),
):
    flag = await _svc.investigate_claim(claim_id, current_user.id, db)
    return {"status": flag.status}


@router.post("/claims/{claim_id}/clear")
async def clear_flag(
    claim_id: UUID,
    body: ClearFlagRequest,
    current_user=Depends(require_auditor),
    db: AsyncSession = Depends(get_db),
):
    flag = await _svc.clear_flag(claim_id, current_user.id, body.resolution_note, db)
    return {"status": flag.status}

File: frontend/src/services/audit.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { api } from './api'
import type { ClaimResponse } from './claims'

export interface AuditFilters {
  violation_type?: string
  date_from?: string
  date_to?: string
  employee_id?: string
}

export function useAuditClaims(filters: AuditFilters) {
  const params = new URLSearchParams()
  if (filters.violation_type) params.set('violation_type', filters.violation_type)
  if (filters.date_from) params.set('date_from', filters.date_from)
  if (filters.date_to) params.set('date_to', filters.date_to)
  if (filters.employee_id) params.set('employee_id', filters.employee_id)
  const query = params.toString()

  return useQuery<ClaimResponse[]>({
    queryKey: ['audit-claims', filters],
    queryFn: () => api.get<ClaimResponse[]>(`/api/v1/audit/claims${query ? `?${query}` : ''}`),
  })
}

export function useInvestigateClaim() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (claimId: string) =>
      api.post<{ status: string }>(`/api/v1/audit/claims/${claimId}/investigate`, {}),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['audit-claims'] }),
  })
}

export function useClearFlag() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ claimId, resolution_note }: { claimId: string; resolution_note: string }) =>
      api.post<{ status: string }>(`/api/v1/audit/claims/${claimId}/clear`, { resolution_note }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['audit-claims'] }),
  })
}
